Draft release notes and features for Gradle 4.8

eriwen opened the pull request:
This is missing a bit of detail and imagery, but I hope will be acceptable for 4.8 RC1

eriwen commented:
@oehme I shamelessly copied on of your comments about incremental annotation processing to a new release notes subsection. Please review and update my additions here.

@lptr I want to say more about immutable tasks. I hope you can help make it clear why this is an important feature and when users should use it.

@eskatos I have not included Kotlin DSL yet here as I don't know the status for 4.8.

Others please chime in here so we can merge this in the next 30 hours or so.

oehme commented:
Thanks for adding the incremental processing Eric!

ijuma commented:
@eriwen Is it right that Java 11 support was added to 4.8-rc1? If so, maybe worth mentioning.

lptr commented:
@eriwen the primary use case is when a plugin wants to register a task and wants to set the value of certain parameters up front, without allowing users to modify it later.

oehme commented:
@ijuma Java 11 support will be in 4.9.

pioterj commented:
@eriwen I think the top two features of this release are dependency locking and stable publishing. I'd mention improvements to incremental annotation processing later in the preamble. I'd also move section about publishing improvements higher in the main body.

For dependency locking, I think we can expand on why is this important. We could mention how this is a superior solution to Maven snapshots. 

For publishing, wouldn't it be good to emphasize more than the new plugins are now stable and recommend users of the old plugins to migrate?

For immutable tasks, I miss some context why it's relevant. Is this what we recommend users to do?

pioterj commented on subprojects/docs/src/docs/release/notes.md:
```diff
@@ -1,3 +1,31 @@
+ 
+The publishing plugins gets some much-needed improvements in this release:
+
+ * The Signing Plugin now supports [signing all artifacts of a publication](#signing-publications).
```
Let's add links to plugin pages in the user guide when mentioning them for the first time.

eriwen replied:
I think such links will distract from folks reading the rest of the notes. I think we should just link to them in the details.

pioterj commented:
@oehme Regarding "TextResources can now be fetched from a URI" it may be worth mentioning this also in the checkstyle and possibly other code quality plugins chapters. We used to have a whole blog post about how to enforce quality standards in multiple projects: http://web.archive.org/web/20150322034610/http://www.gradleware.com/tutorials/feature-spotlight-enforcing-code-quality-standards/

pioterj commented:
@oehme @big-guy Shouldn't we also mention configuration time performance improvements?

ijuma commented:
@oehme Thanks. I was thinking of this fix https://github.com/gradle/gradle/issues/4860, which is the issue I ran into when I tried it for Apache Kafka (which doesn't use Java 11 bytecode btw). I guess that you are probably referring to https://github.com/gradle/gradle/issues/5120, which is additional work that is needed for full Java 11 support.

eriwen commented:
@lptr 

> the primary use case is when a plugin wants to register a task and wants to set the value of certain parameters up front, without allowing users to modify it later.

I think a specific scenario would help users (and me) understand why this is useful and not just "immutability is good because that's what everyone tells me"

eriwen commented:
I have addressed the feedback that I can and will merge this PR. 

@lptr please add more to the immutable tasks details when you can.
@ljacomet could you please expand on the dependency locking details as Hans and Piotr have requested?
@eskatos please add Kotlin DSL 0.17 to the notes when you are able

I will add imagery for 4.8 GA.

eskatos commented:
@eriwen, done in e91fe4d9140624eb41bc724527c2b980cf4cac2f

lptr commented:
See https://github.com/gradle/gradle/pull/5384/

pioterj commented:
@marcphilipp Could you address the feedback regarding stable publishing while @oehme is away? I think we should emphasise more that the new plugins are now stable and recommend users of the old plugins to migrate.

marcphilipp commented:
> The `maven-publish` and `ivy-publish` plugins are now considered stable and the `maven` plugin deprecated — please migrate.

@pioterj Sure, isn't the above clear enough, though?

pioterj commented:
@marcphilipp Sorry, I missed that. I think it wasn't there before. 

As for `maven` plugin being deprecated, I thought we are not doing it in this release yet. I don't see a deprecation notice in the user guide or the command line.

marcphilipp commented:
Yes, you're right, technically it's not deprecated, yet. I figured @eriwen put it like that to emphasize that users should migrate. I'll talk to him and we'll improve it.

pioterj commented:
I think we can tell it like it is - new plugins are recommended over the original ones and the latter will be deprecated in a future release.

marcphilipp commented:
Done in 53937e291200f8d01d62b02fe1d8b5c27b191b45.

@eriwen Feel free to improve the wording.

eriwen commented:
[Good call](https://github.com/gradle/gradle/pull/5342#issuecomment-388863828) @pioterj
